### PR TITLE
Initial applyStashedOp implementation in Shared Array

### PR DESCRIPTION
**Pull request by dannimad**
This PR consists in a first implementation to applyStashedOp in shared array, considering mainly basic ops insert, delete and move. 

Implementation consisted in recreating what the DDS would do while processing each type of operation being treated as remote. The reason for this is because shared array ops are not easily reproducible from the type of metadata they have. For example, map set op contains  the key and the value fields, which is exactly what set parameters are and is not the case for shared array insert op, which needs index and value as well but op contains entryId, value and insertAfterEntryId. That translation is already done in op remote processing so currently just reusing that logic for each operation type. 

Adding a couple e2e tests to verify implementation, being conscious those are the minimum possible testing we could add and once local server stress tests are enabled we could start hitting bugs. 

toggle and toggleMove ops require a bit more nuance to simulate so adding more tests for those in a subsequent PR

### CODE REVIEW

File: packages/dds/legacy-dds/src/array/sharedArray.ts [hunk 1]
```diff
@@ -829,7 +829,76 @@ export class SharedArrayClass<T extends SerializableTypeForSharedArray>
...
+					const isDeleted = liveEntry.isDeleted;
+					this.updateLiveEntry(liveEntry.entryId, newElementEntryId);
+					// mark newly added element as deleted if existing live element was already deleted
+					if (isDeleted) {
+						newElement.isDeleted = isDeleted;
+					}
+				}
+				break;
+			}
+			// case OperationType.toggle: {
```
Comment (Abe27342) [inline packages/dds/legacy-dds/src/array/sharedArray.ts]: remove the commented bits or if you think they are "close to correct", leave a `// TODO: toggle needs something like this for applyStashedOp` or something to that effect above them.

Comment (dannimad) [reply packages/dds/legacy-dds/src/array/sharedArray.ts]: I have the next set of changes enabling toggle and toggleMove ready. If there are no more comments in this PR I would prefer merging as it is and have the next one in line.

File: packages/test/test-end-to-end-tests/src/test/offline/stashedOps.spec.ts [hunk 2]
```diff
@@ -517,6 +533,116 @@ describeCompat("stashed ops", "NoCompat", (getTestObjectProvider, apis) => {
...
+					type: 3,
+					isDeleted: false,
+				} satisfies IToggleOperation);
+				// Attach the revertible event listener.
+				array.on("revertible", (revertibleItem: IRevertible) => {
+					revertible = revertibleItem;
+				});
+				array.insert(0, "test");
+				array.insert(1, "test2");
+				revertible.revert();
```
Comment (ChumpChief) [inline packages/test/test-end-to-end-tests/src/test/offline/stashedOps.spec.ts:589]: Nit: since it's non-obvious to someone who hasn't looked at SharedArray before, might be nice to add a small inline comment here about the relationship between revert/toggle
```suggestion
				// Reverting the revertible generates a toggle op
				revertible.revert();
```

File: packages/test/test-end-to-end-tests/src/test/offline/stashedOps.spec.ts [hunk 3]
```diff
@@ -517,6 +533,116 @@ describeCompat("stashed ops", "NoCompat", (getTestObjectProvider, apis) => {
...
+	it("resends toggle shared array op", async function () {
+		const pendingOps = await generatePendingState(
+			testContainerConfig,
+			provider,
+			false, // Don't send ops from first container instance before closing
+			async (c, d) => {
+				const array = await d.getSharedObject<ISharedArray<string>>(arrayId);
+				let revertible: IRevertible = new SharedArrayRevertible(array, {
+					entryId: "dummy",
+					type: 3,
```
Comment (ChumpChief) [inline packages/test/test-end-to-end-tests/src/test/offline/stashedOps.spec.ts]: Should probably import/use OperationType.toggle here?

File: packages/test/test-end-to-end-tests/src/test/offline/stashedOps.spec.ts [hunk 4]
```diff
@@ -517,6 +533,116 @@ describeCompat("stashed ops", "NoCompat", (getTestObjectProvider, apis) => {
...
+	it("resends toggle move shared array op", async function () {
+		const pendingOps = await generatePendingState(
+			testContainerConfig,
+			provider,
+			false, // Don't send ops from first container instance before closing
+			async (c, d) => {
+				const array = await d.getSharedObject<ISharedArray<string>>(arrayId);
+				let revertible: IRevertible = new SharedArrayRevertible(array, {
+					entryId: "dummy",
+					type: 3,
```
Comment (ChumpChief) [inline packages/test/test-end-to-end-tests/src/test/offline/stashedOps.spec.ts]: Looks like 3 is toggle - should this be 4 (or better, OperationType.toggleMove)?

File: packages/dds/legacy-dds/src/array/sharedArray.ts [hunk 5]
```diff
@@ -829,7 +829,73 @@ export class SharedArrayClass<T extends SerializableTypeForSharedArray>
...
+				if (
+					!this.isLocalPending(op.entryId, "isLocalPendingDelete") &&
+					!this.isLocalPending(op.entryId, "isLocalPendingMove")
+				) {
+					this.updateLiveEntry(this.getLiveEntry(op.entryId).entryId, op.entryId);
+				}
+				break;
+			}
+			default: {
+				throw new Error(`Operation type not supported or recognized: ${op}`);
```
Comment (ChumpChief) [inline packages/dds/legacy-dds/src/array/sharedArray.ts]: Should this be unreachable?  If so, consider using `unreachableCase(op)` instead (e.g. we do this in SharedDirectory's applyStashedOp).

File: packages/dds/legacy-dds/src/array/sharedArray.ts [hunk 6]
```diff
@@ -829,7 +829,73 @@ export class SharedArrayClass<T extends SerializableTypeForSharedArray>
...
 	}
 
-	protected applyStashedOp(_content: unknown): void {
-		throw new Error("Not implemented");
+	protected applyStashedOp(content: unknown): void {
+		const op = content as ISharedArrayOperation<T>;
+
+		switch (op.type) {
+			case OperationType.insertEntry: {
+				this.handleInsertOp<SerializableTypeForSharedArray>(
```
Comment (ChumpChief) [inline packages/dds/legacy-dds/src/array/sharedArray.ts:837]: Overall, it feels not-ideal that there's not a direct mapping of op to normal-ish local method (e.g. like in MapKernel "clear" -> this.clear(), "delete" -> this.delete(), etc.).  Not saying that treating the stashed op as remote won't have the correct ultimate effect necessarily, but it's hard to inspect with confidence and understand the implications.

I don't want to hold this back as a first implementation, but if we find bugs e.g. when adding it to stress/fuzz or more tests, probably should consider refactoring the DDS impl to allow for more-obvious mapping of stashed op to recreation of the local modifications that got it there rather than simulating remote changes.

**Review by ChumpChief — APPROVED** [review]
Approving with comments.  Pending those probably good for a first impl but longer-term will probably want some refactoring.

File: packages/test/test-end-to-end-tests/src/test/offline/stashedOps.spec.ts (hunk 3 above)
Comment (dannimad) [reply packages/test/test-end-to-end-tests/src/test/offline/stashedOps.spec.ts]: these are dummy entries and will be overridden by the listener but makes sense for clarity purposes.

**Review by ChumpChief — APPROVED** [review]
